rllib/algorithms/impala/torch/impala_torch_learner.py

```diff
@@ -0,0 +1,121 @@
+        target_actions_logp = target_policy_dist.logp(batch[SampleBatch.ACTIONS])
+
+        # TODO(Artur): In the old impala code, actions were unsqueezed if they were
+        # multi_discrete. Find out why and if we need to do the same here.
+        # actions = actions if is_multidiscrete else torch.unsqueeze(actions, dim=1)
+
+        target_actions_logp_time_major = make_time_major(
+            target_actions_logp,
+            trajectory_len=self.rollout_frag_or_episode_len,
+            recurrent_seq_len=self.recurrent_seq_len,
+            drop_last=self.vtrace_drop_last_ts,
+        )
+        behaviour_actions_logp_time_major = make_time_major(
+            behaviour_actions_logp,
+            trajectory_len=self.rollout_frag_or_episode_len,
+            recurrent_seq_len=self.recurrent_seq_len,
+            drop_last=self.vtrace_drop_last_ts,
+        )
+        values_time_major = make_time_major(
+            values,
+            trajectory_len=self.rollout_frag_or_episode_len,
+            recurrent_seq_len=self.recurrent_seq_len,
+            drop_last=self.vtrace_drop_last_ts,
+        )
+        bootstrap_value = values_time_major[-1]
+        rewards_time_major = make_time_major(
+            batch[SampleBatch.REWARDS],
+            trajectory_len=self.rollout_frag_or_episode_len,
+            recurrent_seq_len=self.recurrent_seq_len,
+            drop_last=self.vtrace_drop_last_ts,
+        )
+
+        # the discount factor that is used should be gamma except for timesteps where
+        # the episode is terminated. In that case, the discount factor should be 0.
+        discounts_time_major = (
+            1.0
+            - make_time_major(
+                batch[SampleBatch.TERMINATEDS],
+                trajectory_len=self.rollout_frag_or_episode_len,
+                recurrent_seq_len=self.recurrent_seq_len,
+                drop_last=self.vtrace_drop_last_ts,
+            ).type(dtype=torch.float32)
+        ) * self.discount_factor
+
+        # TODO(Artur) Why was there `TorchCategorical if is_multidiscrete else
+        #  dist_class` in the old code torch impala policy?
+        device = behaviour_actions_logp_time_major[0].device
+
+        # TODO(Artur): See if we should compute v-trace corrected targets on CPU
+        vtrace_adjusted_target_values, pg_advantages = vtrace_torch(
+            target_action_log_probs=target_actions_logp_time_major,
+            behaviour_action_log_probs=behaviour_actions_logp_time_major,
+            discounts=discounts_time_major,
+            rewards=rewards_time_major,
+            values=values_time_major,
+            bootstrap_value=bootstrap_value,
+            clip_rho_threshold=self.vtrace_clip_rho_threshold,
+            clip_pg_rho_threshold=self.vtrace_clip_pg_rho_threshold,
+        )
+
+        # Sample size is T x B, where T is the trajectory length and B is the batch size
+        sample_size = convert_to_torch_tensor(
```

Comment:
let's make sure this is the same as the old torch impala code. If there are differences even on the old code, let's leave TODOs here.